### PR TITLE
Redis Services

### PR DESCRIPTION
**Pull request by jtkech**

### CODE REVIEW

**Comment by jtkech** [comment]
@deanmarcussen 

As discussed on gitter you can already try it, i added a Redis Message Bus and made a Distributed Signal that is automatically registered when a Message Bus is registered, so when you enanable Redis Bus, the Signal service becomes distributed.

As discussed on gitter, with the new caching pattern maybe we will not use any message bus, but i put it here (the code is quite simple) to try things with the current caching pattern that uses Signal Tokens. I didn' try it yet with multiple instances, but normally it will keep them in sync, not for all services but most of them as site settings, layers, templates, queries, even they currently use local memory caches.

**Comment by jtkech** [comment]
Just tried with all redis features enabled with 2 instances on the same sql server database

E.g i tried to update the site name on one instance and then see if it is up to date on the front end but by using the other instance. Idem after updating the main menu, a liquid template under the admin.

**Comment by agriffard** [comment]
Is there an implementation of a MessageBus that is planned with RabbitMq?

**Comment by jtkech** [comment]
In fact we will try to not have to use a message bus, but i needed one for testing things with the current caching pattern, that's why i added a redis implementation.

So it's not yet planned but looks like a good idea.

**Comment by deanmarcussen** [comment]
@jtkech I had some time to try this, and it looks good, and I can see the two instances staying in sync.

But I also see and we have skyped about how your distributed branch is altering some of the patterns, so that is interesting too.

One question, if you are able to implement this without a lot of messaging, through the distributed document arrangement on your other pr, will there still be a space for distributed events with data included (so messages)?

I am thinking of the media cache for azure, where I have an old pr https://github.com/OrchardCMS/OrchardCore/pull/5617 to delete removed files from the cache.

That wouldn't work in distributed, and may need more data than a `IChangeToken`, could provide as it would want the name of the file/data on the event.

I'm sure there will be other scenarios where we need some form of distributed event, with data, as well, though nothing comes directly to mind.

**Comment by jtkech** [comment]
@deanmarcussen 

> I had some time to try this, and it looks good, and I can see the two instances staying in sync.

Good to know, thanks for trying it

> will there still be a space for distributed events with data included (so messages)?
> and may need more data than a IChangeToken

This is exactly what i'm now thinking about. In #5249 i started to adapt services to the new pattern, may have been better to do separate PRs but i'm blocked ;)

This pattern allows to keep in sync single Documents of some given types, but as a 1st example there is the `LayerFilter` that checks the `LayerMetadata` change token that doesn't change when we update one of the above Document type, but  through an handler when we update any widget in a layer. So here, right now i kept the use of `_signal.DeferredSignalToken(LayerChangeToken)`.

I'm thinking about other possible solutions and in the above example there is only the change token. Passing data, I already did this in the old distributed branch, e.g for some tenant events with encrypted data if there were not shared. Also indexing commands if using Lucene with local files.

So, this is exactly what i'm now thinking about ;)

Maybe we would have to distributed cache, not only single document types, but also e.g index tables states as the `LayerMetadataIndex`. Maybe some specific Documents kept in sync and that would be, through the cache, a kind of distributed **footprints** that would be used by some services.

File: src/OrchardCore.Modules/OrchardCore.Redis/Manifest.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+using OrchardCore.Modules.Manifest;
+
+[assembly: Module(
+    Name = "Redis",
+    Author = "The Orchard Team",
+    Website = "https://orchardproject.net",
+    Version = "2.0.0"
```
[the remark this answers is not in the document]
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Redis/Manifest.cs:7]: These need to be updated to use the constants.

[the remark this answers is not in the document]
Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.Redis/Manifest.cs:7]: Will do soon, thanks

**Comment by agriffard** [comment]
Even it is now merged, please consider adding some docs about how to use it, configure it, ...

**Comment by agriffard** [comment]
Can we now expect a RabbitMq implementation of the IMessageBus?

**Comment by jtkech** [comment]
@agriffard 

> Can we now expect a RabbitMq implementation of the IMessageBus?

Why not ;) Here it's just a concrete implementation based on redis because currently we need a message bus to invalidate local memory caches, but this will be no more the case with #5249 

> Even it is now merged, please consider adding some docs about how to use it, configure it, ...

Okay, maybe after #5249 because the message bus may not be required, only the redis cache. The redis config is just a configuration string to add e.g. in appsettings.json where i already added a commented example.

**Comment by Piedone** [comment]
There's no mention of this in the docs, right? BTW [here's your demo video](https://www.youtube.com/watch?v=etH6IJOGUe8).

**Comment by jtkech** [comment]
Yes, you're right

But still good to have a doc, i didn't start it waiting for remaining things to do around distributed / stateless configs

- To share and sync documents as layers, templates, queries, site settings ... you just need to enable `Redis.Cache`, but this is not exactly what i demoed, even i used the redis cache feature depending on `Redis`, and showed its config.

- What i demoed is sharing and syncing tenant states, which is another feature `Tenants.Distributed`, but that also needs a concrete implementation of an `IDistributedCache`, so here also `Redis.Cache` is okay.

- I recently did a demo on WF atomicity, i also have pending PRs about background task atomicity, and shell activating atomicity (e.g. for migration steps), to make this working in a distributed env, here we will need a concrete implementation of an `IDistributedLock`, here this is the `Redis.Lock` feature that could be used.

- There is also the `Redis.DataProtection` feature that just needs to be enabled on an existing redis config.

- Recently we made a feature to auto cleanup sitemap cached files (locally) in a distributed env, and that needs a concrete implementation of an `IDistributedCache`, so here also `Redis.Cache` is okay.

- Remaining things to do are related to media cached files, here the number of files may be higher, and currently without any info in a shared document, and finally things related to lucene indexes local files.

**Comment by Piedone** [comment]
Thanks for all the details! I see now, this probably should be wrapped up for docs indeed :).

**Pull request merged**